packages/world/src/EntityId.sol

```diff
@@ -3,21 +3,96 @@ pragma solidity >=0.8.24;
+import { ObjectTypes } from "./ObjectTypes.sol";
+import { ObjectTypeLib } from "./ObjectTypeLib.sol";
+import { Vec3 } from "./Vec3.sol";
+import { MAX_ENTITY_INFLUENCE_HALF_WIDTH } from "./Constants.sol";
+
 type EntityId is bytes32;
 
 library EntityIdLib {
-  function isBaseEntity(EntityId self) internal view returns (bool) {
-    return EntityId.unwrap(BaseEntity._get(self)) == bytes32(0);
+  using ObjectTypeLib for ObjectTypeId;
+
+  function activate(EntityId self) public returns (EnergyData memory) {
+    checkWorldStatus();
+
+    ObjectTypeId objectTypeId = ObjectType.get(self);
+    require(objectTypeId.isActionAllowed(), "Action not allowed");
+
+    requireCallerAllowed(self, objectTypeId);
+
+    // TODO: do we want to support chips for players?
+    EnergyData memory energyData;
+    if (objectTypeId == ObjectTypes.Player) {
+      require(!PlayerStatus._getBedEntityId(self).exists(), "Player is sleeping");
+      energyData = updatePlayerEnergy(self);
+    } else {
+      EntityId forceFieldEntityId;
+      if (objectTypeId == ObjectTypes.ForceField) {
+        forceFieldEntityId = self;
+      } else {
+        (forceFieldEntityId, ) = getForceField(self.getPosition());
+      }
+
+      (energyData, ) = updateMachineEnergy(forceFieldEntityId);
+    }
+
+    require(energyData.energy > 0, "Entity has no energy");
+
+    return energyData;
+  }
+
+  function requireCallerAllowed(EntityId self, ObjectTypeId objectTypeId) internal view {
+    address msgSender = WorldContextConsumerLib._msgSender();
+
+    if (objectTypeId == ObjectTypes.Player) {
+      require(msgSender == ReversePlayer._get(self), "Caller not allowed");
+    } else {
+      address programAddress = self.getProgramAddress();
+      require(msgSender == programAddress, "Caller not allowed");
+    }
+  }
+
+  function requireCallerAllowed(EntityId self) internal view {
+    ObjectTypeId objectTypeId = ObjectType.get(self);
+    requireCallerAllowed(self, objectTypeId);
   }
 
   function baseEntityId(EntityId self) internal view returns (EntityId) {
     EntityId base = BaseEntity._get(self);
-    return EntityId.unwrap(base) == bytes32(0) ? self : base;
+    return EntityId.unwrap(base) == 0 ? self : base;
+  }
+
+  // TODO: add pipe connections
+  // TODO: should non-player entities have a range > 1?
+  function requireConnected(EntityId self, EntityId other) internal view returns (Vec3, Vec3) {
+    Vec3 otherCoord = other.getPosition();
+    return requireConnected(self, otherCoord);
+  }
+
+  function requireConnected(EntityId self, Vec3 otherCoord) internal view returns (Vec3, Vec3) {
+    Vec3 selfCoord = self.getPosition();
+    require(selfCoord.inSurroundingCube(otherCoord, MAX_ENTITY_INFLUENCE_HALF_WIDTH), "Entity is too far");
+    return (selfCoord, otherCoord);
+  }
+
+  function getPosition(EntityId self) internal view returns (Vec3) {
+    return ObjectType._get(self).isMovable() ? MovablePosition._get(self) : Position._get(self);
```

Comment:
In one tx ._get might be called in multiple places, eg in activate its called to check action allowed and then for position. Is there a way to somehow cache this and not do the table read (might not be possible in EVM/Solidity just a though)

Reply:
yea the code is easier to understand when you're not passing around stuff

Reply:
> could be stored in transient storage
wdym?

Reply:
like when reading from tables you could cache them in transient storage which is cheaper than storage, but not sure if the overhead would be worth it in most cases

Reply:
ahh neat, yea sounds like an optional MUD thing if ever implemented